lib/bots/notifybot.ts

```diff
@@ -0,0 +1,918 @@
+	 * @returns          An updated component information object.
+	 */
+	private getNewPRs(component: ComponentTracker): Promise<ComponentTracker> {
+		// Get the changelog for the component from its `master`, as this file
+		// *should* include all given versions. If it doesn't, then something is broken
+		// or the keyframe is incorrect.
+		const repo = this.getRepoDetails(component.repo);
+		const validVersion = (data: string, version: string) => {
+			return data.indexOf(version) !== -1;
+		};
+
+		if (!repo) {
+			throw new Error(`Cannot find appropriate repo/owner for ${component.repo}`);
+		}
+
+		/** Retrieve the changelog. */
+		const fileRequest = {
+			api: this.githubApi,
+			filepath: ChangelogFile,
+			hash: 'HEAD',
+			owner: repo.owner,
+			repo: repo.repo,
+		};
+		return this.retrieveFileFromHash(fileRequest).then((file) => {
+			if (!file) {
+				throw new Error(`Couldn't find the CHANGELOG.md in ${repo.owner}/${repo.repo}`);
+			}
+
+			// We have the Changelog, we want to cull any versions which aren't
+			// relevant by removing them. Do this by splitting the lines based on
+			// the old version and the new version.
+			// For both incrementing and decrementing versions, we want all the versions
+			// that exist between the greatest version (inclusive) and the oldest version
+			// (exclusive).
+			// We can easily get this by swapping the versions for decrementing.
+			let changeParts = [ file ];
+			let changelogString = '';
+
+			// Ensure that the old version (if there is one) and the new version are valid.
+			if (component.oldVersion) {
+				if (!validVersion(file, component.oldVersion)) {
+					this.reportError(new NotifyBotError(NotifyErrorCodes.OldVersionInvalid,
+						`The old version of the component in the old keyframe was invalid`,
+						repo.owner,
+						repo.repo,
+						component.oldVersion
+					));
+					return component;
+				}
+			}
+			if (!validVersion(file, component.newVersion)) {
+				this.reportError(new NotifyBotError(NotifyErrorCodes.NewVersionInvalid,
+					`The new version of the component in the new keyframe was invalid`,
+					repo.owner,
+					repo.repo,
+					component.newVersion
+				));
+
+				return component;
+			}
+
+			// If there's no old version, then *anything* we release is essentially new and
+			// therefore is now implemented. It is implicitly an incrementing version.
+			if (!component.oldVersion) {
+				component.incrementing = true;
+			} else if (component.oldVersion > component.newVersion) {
+				// A new version less than an older version is a version down.
+				component.incrementing = false;
+			} else if (component.newVersion > component.oldVersion) {
+				component.incrementing = true;
+			}
+
+			// If there is an old component, use it to cull anything before that version.
+			if (component.oldVersion) {
+				changeParts = changeParts[0].split(component.incrementing ? component.oldVersion :
+					component.newVersion);
```

Comment:
> We're going to go ahead and (at some point) go for a module that parses changelogs based.

...based on what? 

Interesting thought that occurs to me: why bother parsing the changelog itself? We can get close with regexes here, but it's not perfect - it's a plain text human readable document with all the fuzzy mess that entails. How hard would it be get the commits + messages directly, and just pull this data from them? There you've got much cleaner & more structured data. I can imagine in some places that might be tricky to quickly grab though...

Could easily start getting versionbot to add hidden metadata in changelogs to make this easier instead? You should be able to put in HTML comments with arbitrary data inline, that don't actually appear in the human readable doc, and are _very_ unlikely to conflict with : `<!--VB { version: '1.2.3', pr: '#123', user: 'pimterry' } -->`

<!-- yes, like this! -->

Alternatively, there's a way more long winded approach where we treat human readable changelogs as a final artefact, but generate a machine readable one first, so we have both available.